### PR TITLE
Write example dynamic parameters generated code

### PR DESCRIPTION
**Pull request by jsn0716**
### Description
Write example generated dynamic parameter code. After this is approved (merged?), will write the actual generation code such that the generated code is the same as this manually written code.

### Testing Done
No testing, just example code. Testing will be done with actual generated code.

### Resolved Issues
Part of #1447

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: you can make use of `allValuesRefereeCommand()` to avoid parsing the c++ code, and use `fromStringToRefereeCommand()` for the default value.

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: we need to also add the `example_factory_param` to see how that looks

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Thoughts on generating enum values as `boost::program_options::value<std::string>` and then do `this->mutableExampleEnumParam()->setValue(fromStringToRefereeCommand(args.example_enum_param))`?

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: the included configs would also have to get tossed in here, (with the respective boost cli args generated) we might want to share configs between different binaries

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: should we append Config to the end? `fooConfig` vs `foo`

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Nice work :100:, just a few comments

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: If we'll just use all the values, do we even need that argument for `EnumeratedParameter`?

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Can you give an example of how configs might be shared between binaries? Also, not sure if I completely understand what you mean by with the respective cli args generated.

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (amrm3lm) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: if this is to be merged, a brief comment explaining the document would be helpful

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (amrm3lm) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: add eof

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (amrm3lm) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: are we still planning on implementing optional values? if so should we have an example of it?

**Review by amrm3lm — COMMENTED** [review]
lgtm, just a couple of small things and a question

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jonathanlew) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:36]: just checking that it isn't a style convention to name parameters using the type

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jonathanlew) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Usually allValues should be what we want so the argument might be unnecessary. Maybe the argument to EnumeratedParameter should be allValuesRefereeCommand when a user instantiates it? @akhilveeraghanta

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jonathanlew) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:328]: what is this used for?

**Review by jonathanlew — COMMENTED** [review]
looking good

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jonathanlew) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: would be good to have a rough idea of what the intended input was, i.e. what configs include other configs, etc.

Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: its from the `shared/parameter_v2/config_definitions/test` folder

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:36]: nope its not, the example params just have the type in the name

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: `EnumeratedParameters` aren't necessarily enums, they can be strings with a finite set of values, i think the factory one would be an `EnumeratedParameter`

That being said, not sure if `EnumeratedParameter` is the _most clear_ name

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: `parameter/config/full_system_main_command_line_args.yaml`
`parameter/config/standalone_simulator_main_command_line_args.yaml`
`parameter/config/robot_diagnostics_main_command_line_args.yaml`

All share the networking parameters, this is a small example, but we could put networking stuff in its own config file, and then include it in all the `***main_command_line_args.yaml` to avoid duplicating the definitions like we do now.

If we included the networking config into the example config, the struct would be generated to something like this
~~~
  struct commandLineArgs {
            bool help = false;
            bool example_bool_param = true;
            int example_int_param = 3;
            uint example_uint_param = 3;
            float example_float_param = 4.04;
            // this would be the included config
           struct NetworkingConfig {
                std::string interface;
           }
~~~

and we would have the corresponding boost cli stuff generated
~~~ 
desc.add_options()("interface", boost::program_options::value<std::string>(
&args.NetworkingConfig.interface), "network interface");
~~~

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: Yeah once this is actually generated, not sure if this is something people should be reading directly vs just the config files

Comment (jonathanlew) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: Can we just add a quick comment at the top since it isn't obvious

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Then it seems like we need a rule for these config files that any config parameter names have to be unique across all configs, or at least if any of them have dependencies.

Another option could be to prepend the config name: 
`desc.add_options()("networkingConfig.interface", boost::program_options::value<std::string>(
&args.NetworkingConfig.interface), "network interface");`

Tried the second approach, let me know what you think.

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:17]: Done

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Done, added as part of `exampleConfig`

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Done

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Done

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Done

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Sounds good, done

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: yeah it would need that info, I think currently we have the `cpp_dynamic_parameter` rule defined in `//shared/parameter_v2/parameter.bzl` which relays the "enum_deps" information to the generation scripts. That might have the file paths but youd need to double check, ideally we can get that info through Bazel.

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Yeah that makes a lot of sense, there is no requirement right now that parameters have to have unique names across multiple configs, so that looks good

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (akhilveeraghanta) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp:328]: yeah its used by the visualizer to recursively visit over the tree and generate sliders/buttons

**Review by akhilveeraghanta — COMMENTED** [review]
Noice this is almost ready to merge :fire:, could you also stage the Bazel BUILD file so we know it compiles

**Comment by akhilveeraghanta** [comment]
@jsn0716 throwing a curve ball your way: could you also think about how we could register callbacks for the entire config https://github.com/UBC-Thunderbots/Software/issues/1550

**Comment by jsn0716** [comment]
> 
> 
> @jsn0716 throwing a curve ball your way: could you also think about how we could register callbacks for the entire config #1550

Sure I can look into it, can I get a code pointer to an example of manual registration mentioned in #1550

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jonathanlew) [inline src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: there's some discussion on slack about representing enums as strings for dynamic params. I think it's relevant here

**Comment by jsn0716** [comment]
Made the changes necessary for this example file to build. I added a `BUILD` file as well as a dummy test file, since bazel seems to ignore the `cc_library` unless there is a dependence on it. I will remove those two files before merging.

This change includes adding another bazel option to get optional values working; for example, `--example_optional_param` and `--example_optional_param_nullopt`, where the latter needs to be set to `true` to assign `nullopt` to `example_optional_param`.

Another change is the ugly manual addition of different types to `MutableParameterVariant` and `ParameterVariant` in `config.h`. Doing this manually for now, but I'm hoping that eventually this will just be generated as well.

Lastly, `NumericParameter` doesn't work with `std::optional` at the moment, but I think this can just be addressed after we have most features working.

**Comment by jsn0716** [comment]
Removed std::optional, changed enum types to string, and removed temp BUILD / test files

File: src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp
Comment (jsn0716) [reply src/shared/parameter_v2/examples/generated_dynamic_parameters.hpp]: Made changes as discussed

**Comment by akhilveeraghanta** [comment]
lgtm pending CI!

**Review by jonathanlew — COMMENTED** [review]
LGTM

File: src/software/parameter/config.h
Comment (jonathanlew) [inline src/software/parameter/config.h]: nit: this probably isn't needed

Comment (jsn0716) [reply src/software/parameter/config.h]: Removed

**Review by jonathanlew — CHANGES_REQUESTED** [review]
It turns out that the game_state.h include is needed. :( How would we know this though? Can we add a field in the example.yaml so we know to include it? Thoughts? @akhilveeraghanta @jsn0716

**Comment by akhilveeraghanta** [comment]
> It turns out that the game_state.h include is needed. :( How would we know this thought? Can we add a field in the example.yaml so we know to include it? Thoughts? @akhilveeraghanta @jsn0716

The build is failing on the dynamic parameter widget, the variant visitor is still using the Parameter<int> instead of NumericParameter<int> 

dynamic param widget needs to be update to the new `MutableParamVariant` def

**Comment by jsn0716** [comment]
Build passes now

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
:100:

**Pull request merged**
